### PR TITLE
Move more semantic functions from func.d to funcsem.d and typesem.d

### PR DESCRIPTION
**Pull request by muscar**

### CODE REVIEW

File: compiler/src/dmd/cxxfrontend.d [hunk 1]
```diff
@@ -265,10 +265,16 @@ bool functionSemantic3(FuncDeclaration fd)
...
     import dmd.funcsem;
-    return dmd.funcsem.leastAsSpecialized(f, g, names);
+    return dmd.funcsem.leastAsSpecialized(fd, g, names);
+}
+
+PURE isPure(FuncDeclaration fd)
+{
+    import dmd.funcsem;
+    return dmd.funcsem.isPure(fd);
 }
```
Comment (thewilsonator) [inline compiler/src/dmd/cxxfrontend.d:278]: I don't think we need to expose this to C++, it is unused by LDC

File: compiler/src/dmd/funcsem.d [hunk 2]
```diff
@@ -3062,3 +3062,361 @@ extern (D) bool checkNRVO(FuncDeclaration fd)
...
+                if (!fd.isTypeIsolated(tmi, parentTypes))
+                    return false;
+            }
+            return true;
+
+        default:
+            return true;
+    }
+}
+
```
Comment (thewilsonator) [inline compiler/src/dmd/funcsem.d:3422]: and I think the style checker requires that this trailing newline be removed.

File: compiler/src/dmd/cxxfrontend.d (hunk 1 above)
Comment (muscar) [reply compiler/src/dmd/cxxfrontend.d:278]: @thewilsonator thanks. Any way I can check before I make furure PRs?

Comment (thewilsonator) [reply compiler/src/dmd/cxxfrontend.d:278]: I use `git grep` on my local clone of LDC. I guess you could also use GitHub search and use the language filter and see if you get any hits of C++ files.

**Comment by thewilsonator** [comment]
You will also need to update `frontend.h`

```diff
--- a/home/circleci/dmd/compiler/src/dmd/frontend.h
+++ b/home/circleci/dmd/generated/linux/release/64/frontend.h
@@ -3621,14 +3621,6 @@ struct ObjcFuncDeclaration final
         {}
 };
 
-enum class PURE : uint8_t
-{
-    impure = 0u,
-    fwdref = 1u,
-    weak = 2u,
-    const_ = 3u,
-};
-
 enum class VarArg : uint8_t
 {
     none = 0u,
@@ -4364,6 +4356,14 @@ enum class TRUST : uint8_t
     safe = 3u,
 };
 
+enum class PURE : uint8_t
+{
+    impure = 0u,
+    fwdref = 1u,
+    weak = 2u,
+    const_ = 3u,
+};
+
 class TypeFunction final : public TypeNext
 {
 public:
```

File: compiler/src/dmd/cxxfrontend.d (hunk 1 above)
Comment (ibuclaw) [reply compiler/src/dmd/cxxfrontend.d:278]: LDC isn't the only compiler, so don't make presumptions from one data point. ;-)

Comment (thewilsonator) [reply compiler/src/dmd/cxxfrontend.d:278]: for codegen functions yes, but for semantic functions its a pretty good barometer

**Review by thewilsonator — APPROVED** [review]

File: compiler/src/dmd/cxxfrontend.d [hunk 3]
```diff
@@ -265,10 +265,10 @@ bool functionSemantic3(FuncDeclaration fd)
     return dmd.funcsem.functionSemantic3(fd);
 }
 
-MATCH leastAsSpecialized(FuncDeclaration f, FuncDeclaration g, Identifiers* names)
+MATCH leastAsSpecialized(FuncDeclaration fd, FuncDeclaration g, Identifiers* names)
 {
     import dmd.funcsem;
-    return dmd.funcsem.leastAsSpecialized(f, g, names);
+    return dmd.funcsem.leastAsSpecialized(fd, g, names);
 }
```
Comment (ibuclaw) [inline compiler/src/dmd/cxxfrontend.d:272]: Unrelated change?

File: compiler/src/dmd/declaration.h [hunk 4]
```diff
@@ -35,6 +35,7 @@ namespace dmd
     bool functionSemantic(FuncDeclaration* fd);
     bool functionSemantic3(FuncDeclaration* fd);
     MATCH leastAsSpecialized(FuncDeclaration *f, FuncDeclaration *g, Identifiers *names);
+    PURE isPure(FuncDeclaration *f);
```
Comment (ibuclaw) [inline compiler/src/dmd/declaration.h:38]: This is missing from cxxfrontend -> will cause undefined symbol errors.

**Review by ibuclaw — CHANGES_REQUESTED** [review]
